lib/spack/spack/solver/asp.py

```diff
@@ -878,6 +878,7 @@ def key_fn(version):
             if (
                 isinstance(v.version, spack.version.GitVersion)
                 and v.version.user_supplied_reference
+                and v.origin != version_provenance.packages_yaml
```

Comment:
Without this, I cannot set `version_provenance.packages_yaml` below like:

```
self.add_concrete_versions_from_specs(req_version_specs, version_provenance.packages_yaml)
```

While I think this change is technically correct, I wonder if the uses of `hash=numbered-version` in `packages.yaml` is confusing (it is to me).